kubelet: fix some race conditions with logs -f

giuseppe opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

fixes some race conditions with `kubectl logs -f`.  More details are in the message for each commit.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

giuseppe committed:
kubelet: read immediately after creating the watcher

if some events happen between the Read and while we set the watcher,
we might miss them.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

giuseppe committed:
kubelet: handle recreated log files

if the runtime is configured to rotate the log file, we might end up
watching the old fd where there are no more writes.

When a fsnotify event other than Write is received, reopen the log
file and recreate the watcher.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

giuseppe committed:
kubelet: force checking the log file every second

it seems fsnotify can miss some read events, blocking the kubelet to
receive more data from the log file.

If we end up waiting for events with fsnotify, force a read from the
log file every second so that are sure to not miss new data for longer
than that.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

mrunalp commented:
@feiskyer @Random-Liu ptal. This solves issues we have been seeing in our tests for a while.

feiskyer reviewed (commented):
/lgtm

feiskyer commented:
/ok-to-test

wking commented on pkg/kubelet/kuberuntime/logs/logs.go:
```diff
@@ -373,34 +389,40 @@ func isContainerRunning(id string, r internalapi.RuntimeService) (bool, error) {
+			case fsnotify.Create:
+			case fsnotify.Rename:
+			case fsnotify.Remove:
+			case fsnotify.Chmod:
```
I don't know what I'm doing, so ignore me if I make no sense, but it's surprising to me that `Chmod` counts as "recreated".  Should we silently ignore it instead?

giuseppe replied:
I've added it as I've seen the Chmod event coming from fsnotify instead of Rename or Create when the log file was replaced with rename(2).

wking commented on pkg/kubelet/kuberuntime/logs/logs.go:
```diff
@@ -373,34 +389,40 @@ func isContainerRunning(id string, r internalapi.RuntimeService) (bool, error) {
 			case fsnotify.Write:
-				return true, nil
+				return true, false, nil
+			case fsnotify.Create:
```
Don't you need [explicit `fallthrough` statements][1] in Go?  https://play.golang.org/p/1nol1bGX2DR (or maybe I'm just misunderstanding your intention here?)

giuseppe replied:
yes I do :-)

giuseppe commented:
/test pull-kubernetes-local-e2e-containerized

wking commented:
All green :).

mrunalp commented:
/lgtm

giuseppe committed:
logs: consume all file until EOF on exited container

If the container is not found, do not stop reading the log file
immediately but wait until we reach again EOF.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

giuseppe commented:
/test pull-kubernetes-local-e2e-containerized

giuseppe commented:
/test pull-kubernetes-local-e2e-containerized

giuseppe commented:
I've added a patch but I don't think the CI failures depend from it.  Could someone please take a look?

giuseppe commented:
/test pull-kubernetes-local-e2e-containerized

tnozicka commented:
that job is broken on other PRs as well - https://gubernator.k8s.io/builds/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-local-e2e-containerized

maybe that's an optional job by prow, you'll find out from tide status once your PR is tagged
/skip

giuseppe commented:
@mrunalp @feiskyer could you please take a look at the last patch I've added to this PR?

feiskyer commented:
lgtm, could you squash the commits?

/retest

giuseppe commented:
/test pull-kubernetes-local-e2e-containerized

giuseppe commented:
/test pull-kubernetes-local-e2e-containerized

giuseppe commented:
/test pull-kubernetes-local-e2e-containerized

tnozicka commented:
`pull-kubernetes-local-e2e-containerized`  is almost permanently failing across PRs, you don't need to get it passing (probably don't have a chance to do so), just do /skip - it is optional https://github.com/kubernetes/test-infra/blob/bd245500c9bcdbab0b16348a15342892bfdf462e/config/jobs/kubernetes/sig-node/local-e2e-containerized.yaml#L8

/skip

giuseppe commented:
@feiskyer could you please take a look?

mrunalp commented:
/skip

mrunalp commented:
/lgtm

mrunalp commented:
cc: @smarterclayton for approval

smarterclayton commented:
How do we add an e2e that can prevent regressions that led to this?

giuseppe commented:
> How do we add an e2e that can prevent regressions that led to this?

I am not really sure this is enough to fix the issue completely.  By inspecting the code I don't see other possible race conditions and https://github.com/openshift/origin/pull/21795 gives me some confidence about it.  The key to trigger the race conditions seems very high load and/or being throttled.

I'll take a look at existing e2e tests to see if there is anything that can help

smarterclayton commented:
How about an e2e test like the one in openshift that triggered this?  Short
running pods that do log -f on other pods, which you can then scrape via
log and verify all the output was captured?

On Thu, Feb 7, 2019 at 2:06 AM Giuseppe Scrivano <notifications@github.com>
wrote:

> How do we add an e2e that can prevent regressions that led to this?
>
> I am not really sure this is enough to fix the issue completely. By
> inspecting the code I don't see other possible race conditions and
> openshift/origin#21795 <https://github.com/openshift/origin/pull/21795>
> gives me some confidence about it. The key to trigger the race conditions
> seems very high load and/or being throttled.
>
> I'll take a look at existing e2e tests to see if there is anything that
> can help
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/73041#issuecomment-461309808>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p6zlzSLUqh48zBD_GYbD_Dpa4Ulqks5vK9BmgaJpZM4aGnlM>
> .
>

tnozicka commented:
like bringing this one upstream? https://github.com/openshift/origin/pull/20933/files

giuseppe commented:
@smarterclayton would that be ok for upstream?

giuseppe commented:
@smarterclayton, is it fine to merge this PR and successively add upstream the test @tnozicka pointed out?

giuseppe commented:
should this be in v1.14?

giuseppe commented:
added a test to stress "logs -f" not missing the last part of the output

giuseppe commented:
is there anything more I could do to get this moving?

mrunalp commented:
/retest

smarterclayton commented:
/lgtm
/approve

giuseppe committed:
test, e2e: add tests for logs -f

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

mrunalp commented:
/lgtm

mrunalp commented:
/test pull-kubernetes-e2e-gce
